### PR TITLE
Add a Forecasting tutorial on state space models

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2311 

See notebook via [nbviewer](https://nbviewer.jupyter.org/github/pyro-ppl/pyro/blob/forecast-tutorial-ii/tutorial/source/forecasting_ii.ipynb).

### CODE REVIEW

**Comment by fritzo** [comment]
@martinjankowiak @fehiepsi I'm currently hitting numerical issues as in @martinjankowiak 's #2342 ...

@martinjankowiak At some point it would be nice to add a tutorial on pyro.contrib.timeseries. Maybe after this merges you could add a couple sections to this tutorial. No rush though.

**Comment by martinjankowiak** [comment]
another great tutorial!

nits/comments/suggestions/questions:
- [ ] summary section: "use GaussianHMM noise_dist" add "a"
- [ ] "through regression and local level" missing a word?
- [ ] "thus both methods allow parallelization of long time series analysis _even for univariate data_." what do you have in mind?
- [ ] for `Model1` comment on unsqueezing of `trans_matrix` and value of `hidden_dim`?
- [ ] 'clean gradients' => low variance?
- [ ] typo: 'as we have ont'
- [ ] "# apply domain knowledge: the samples must be positive" why not log transform targets?
- [ ] mark xmas on plot?
- [ ] " This is the only change." => " This is the only change from Model1"
- [ ] typo "gradients are are noisier "
- [ ] is there any motivation for sharing `stability` besides simplicity?

File: pyro/distributions/stable.py [hunk 1]
```diff
@@ -10,11 +10,17 @@
...
-def _unsafe_standard_stable(alpha, beta, V, W, coords):
+def _unsafe_standard_stable(alpha, beta, V, W, coords, tol=1e-6):
     # Implements a noisily reparametrized version of the sampler
     # Chambers-Mallows-Stuck method as corrected by Weron [1,3] and simplified
     # by Nolan [4]. This will fail if alpha is close to 1.
 
+    # Numerically stabilize.
+    if tol:
+        assert 0 < tol < 1
+        V = V.mul(1 - tol)
```
[the remark this answers is not in the document]
Comment (fritzo) [reply pyro/distributions/stable.py]: Clamping discards gradients near the edge. I'd rather preserve gradients if we can.

[the remark this answers is not in the document]
Comment (martinjankowiak) [reply pyro/distributions/stable.py]: **update**: the `tol` fix works for my example (i.e. my code runs to completion) for `tol=1.0e-4` and `tol=1.0e-5` but i still get NaNs for `tol=1.0e-6`

**Comment by fritzo** [comment]
@martin thanks for your detailed review!

> "# apply domain knowledge: the samples must be positive" why not log transform targets?

Indeed I have`log1p`-transformed the data. However the data is integer and often zero, and hence its `log1p` transform is positive.

> is there any motivation for sharing stability besides simplicity?

I have two motivations: (1) HMMs with common stability factor are analytic (their conditional marginals are Stable) and hence a parsimonious well-studied model class. (2) the stability parameter needs lots of data to train, so by sharing it across multiple model components we can share statistical strength. However you're right that we could separate the two stability parameters for a more flexible model.

**Comment by fritzo** [comment]
@fehiepsi I'll try your fix from #2344

File: tutorial/source/forecasting_ii.ipynb [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+    "\n",
+    "Pyro implements a few state space models, but the most important are the [GaussianHMM](http://docs.pyro.ai/en/latest/distributions.html#gaussianhmm) distribution and its heavy-tailed generalization the [LinearHMM](http://docs.pyro.ai/en/latest/distributions.html#linearhmm) distribution. Both of these model a linear dynamical system with hidden state; both are multivariate, and both allow learning of all process parameters. On top of these the [pyro.contrib.timeseries](http://docs.pyro.ai/en/latest/contrib.timeseries.html) module implements a variety of multivariate Gaussian Process models that compile down to `GaussianHMM`s.\n",
+    "\n",
+    "Pyro's inference for `GaussianHMM` uses parallel-scan Kalman filtering, allowing fast analysis of very long time series. Similarly, Pyro's inference for `LinearHMM` uses entirely parallel auxiliary variable methods to reduce to a `GaussianHMM`, which then permits parallel-scan inference. Thus both methods allow parallelization of long time series analysis, even for a single univariate time series.\n",
+    "\n",
+    "Let's again look at the [BART train](https://www.bart.gov/about/reports/ridership) ridersip dataset:"
```
Comment (fehiepsi) [inline tutorial/source/forecasting_ii.ipynb]: typo: ridership

**Pull request merged**